drivers: i2c: stm32: add timeout to avoid infinite loop

73jn opened the pull request:
Fixes issue where STM32 I2C LL driver can get stuck in an infinite loop when SDA and SCL lines are short-circuited and interrupts are disabled (CONFIG_I2C_STM32_INTERRUPT=n).

Added timeout mechanism to all blocking wait loops in the STM32 LL I2C driver to avoid indefinite blocking.

Fixes #88506

73jn commented on drivers/i2c/Kconfig.stm32:
```diff
@@ -64,4 +64,12 @@ config I2C_STM32_V2_DMA
+config STM32_I2C_TRANSFER_TIMEOUT_MSEC
+	int "I2C transfer timeout in milliseconds"
+	default 500
+	depends on I2C_STM32_V1 || I2C_STM32_V2
```
Hi @erwango, you're right, that's actually superfluous. Unfortunately I wanted to rebase on MAIN and that closed the PR... is it possible to reopen it?

GeorgeCGV commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -933,24 +945,51 @@ static int stm32_i2c_msg_write(const struct device *dev, struct i2c_msg *msg,
 	}
 
-	return msg_done(dev, msg->flags);
+	ret = msg_done(dev, msg->flags);
+	if (ret < 0) {
+		goto error;
+	}
```
It didn't disable the peripheral when the `msg_done` returns an error. Also, it is not done in `stm32_i2c_msg_read` for the `msg_done` error.

73jn replied:
@erwango I'm going to do it later today

kartben commented:
@erwango can you please help gather the additional approval required to merge this?

GeorgeCGV commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -1302,5 +1323,21 @@ int i2c_stm32_transaction(const struct device *dev,
+		}
+	#endif
+		return -EIO;
+	}
 	return ret;
```
```suggestion
	}
#endif /* !CONFIG_I2C_STM32_INTERRUPT */

	return ret;
```

73jn commented:
@GeorgeCGV Thanks, it's really cleaner. I wanted to follow the way the interrupted part was done so as not to change the behaviour, but obviously it wasn't as good.

GeorgeCGV reviewed (dismissed):
LGTM, thank you

73jn commented:
rebase on main

Commit:
drivers: i2c: stm32: add timeout to avoid infinite loop

Fix issue where STM32 I2C LL driver could block forever when SDA and SCL
are shorted and interrupts are disabled (CONFIG_I2C_STM32_INTERRUPT=n).

Added timeouts to all blocking wait loops in the STM32 LL I2C driver to
avoid indefinite blocking.

Fixes #88506

Signed-off-by: Jean Nanchen <jean.nanchen@hevs.ch>

73jn commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -1302,5 +1322,25 @@ int i2c_stm32_transaction(const struct device *dev,
+		if (LL_I2C_IsEnabledReloadMode(i2c)) {
+			LL_I2C_DisableReloadMode(i2c);
+		}
+	#if defined(CONFIG_I2C_TARGET)
```
done! thanks

73jn commented on drivers/i2c/i2c_ll_stm32_v2.c:
```diff
@@ -1302,5 +1322,26 @@ int i2c_stm32_transaction(const struct device *dev,
 	} while (rest > 0U);
 
+#ifndef CONFIG_I2C_STM32_INTERRUPT
+	struct i2c_stm32_data *data = dev->data;
```
Hi @erwango @GeorgeCGV @etienne-lms, two variables are missing here:
```
	struct i2c_stm32_data *data = dev->data;
	const struct i2c_stm32_config *cfg = dev->config;
	I2C_TypeDef *i2c = cfg->i2c;
```
What is the procedure to follow?

etienne-lms replied:
Indeed we missed that :-\
How come CI build test did not found this issue.

Can you create a pull request to get this fixed?
